src/BoundsInference.cpp

```diff
@@ -109,27 +110,178 @@ class BoundsInference : public IRMutator {
     set<string> in_pipeline, inner_productions;
     Scope<int> in_stages;
 
+    struct CondValue {
+        Expr cond; // Condition on params only (can't depend on loop variable)
+        Expr value;
+
+        CondValue(const Expr &c, const Expr &v) : cond(c), value(v) {}
+    };
+
     struct Stage {
         Function func;
         size_t stage; // 0 is the pure definition, 1 is the first update
         string name;
         vector<int> consumers;
         map<pair<string, int>, Box> bounds;
-        vector<Expr> exprs;
+        vector<CondValue> exprs;
+        set<ReductionDomain, ReductionDomain::Compare> rdoms;
         string stage_prefix;
 
-        // Computed expressions on the left and right-hand sides
+        // Computed expressions on the left and right-hand sides.
+        // Note that a function definition might have different LHS or reduction domain
+        // (if it's an update def) or RHS per specialization. All specializations
+        // of an init definition should have the same LHS.
+        // This also pushes all the reduction domains it encounters into the 'rdoms'
+        // set for later use.
+        pair<vector<CondValue>, vector<CondValue>> compute_exprs_helper(const Definition& def, bool is_update) {
+            internal_assert((!is_update && !def.domain().defined()) || is_update)
+                << "Init definition shouldn't have a RDom\n";
+
+            vector<CondValue> values;
+            vector<CondValue> args;
+
+            // Default case (no specialization)
+            vector<Expr> predicates;
+            if (def.domain().defined()) {
+                rdoms.insert(def.domain());
+                predicates = def.domain().split_predicate();
+            }
+            for (const Expr &val : def.values()) {
```

Comment:
This might remove repeated code in the rest of this block:

    vector<CondValue> vecs = {def.values()};
    if (is_update) {vecs.push_back(def.args());}

    ...